### PR TITLE
Parse CI Artifacts

### PR DESCRIPTION
**Pull request by scbedd**
FYI @JimSuplizio

This is what a result looks like:

```jsonc
{
  "Name": "spring-cloud-azure-starter-monitor",
  "Version": "1.0.0-beta.7",
  "DevVersion": null,
  "DirectoryPath": "sdk/spring/spring-cloud-azure-starter-monitor",
  "ServiceDirectory": "spring",
  "ReadMePath": "sdk/spring/spring-cloud-azure-starter-monitor/README.md",
  "ChangeLogPath": "sdk/spring/spring-cloud-azure-starter-monitor/CHANGELOG.md",
  "Group": "com.azure.spring",
  "SdkType": "spring",
  "IsNewSdk": true,
  "ArtifactName": "spring-cloud-azure-starter-monitor",
  "ReleaseStatus": "Unreleased",
  "IncludedForValidation": false,
  "AdditionalValidationPackages": null,
  // pulled directly from ci.*.yml
  "ArtifactDetails": {
    "name": "spring-cloud-azure-starter-monitor",
    "groupId": "com.azure.spring",
    "safeName": "springcloudazurestartermonitor",
    "skipPublishDocGithubIo": true,
    "skipPublishDocMs": true,
    "skipUpdatePackageJson": true,
    "skipVerifyChangelog": true,
    "releaseInBatch": true
  }
}

```

Which includes pulling from a [`ci.blah.yml`](https://github.com/Azure/azure-sdk-for-java/blob/d9093c6efbadad97a53b22b5eb3d2a00feb66e5b/sdk/spring/ci.monitor.yml#L33)

Confirmed unbroken package-properties on:

- [x] android (new unrelated failure in `main` though)
- [x] ios (no package-properties usage)
- [x] go (no artifacts to save to the packageinfo.json, but no crash either)
- [x] net
- [x] python
- [x] java
- [x] js
- [x] c (doesn't use Package-Properties)
- [x] cpp

### CODE REVIEW

File: eng/common/scripts/Package-Properties.ps1
Comment (scbedd) [inline eng/common/scripts/Package-Properties.ps1:122]: Before someone says

> you forgot to add `hidden` to this one too Scott

This is because `azure-sdk-for-js` sets `artifactName` _after_ they create the package prop class instance for the first time, so their `Get-PackagePropertiesFromRepo` needs to _call_ `InitializeCIArtifact` to populate the ArtifactDetails field.

File: eng/common/scripts/Helpers/Package-Helpers.ps1
Comment (scbedd) [inline eng/common/scripts/Helpers/Package-Helpers.ps1:77]: On second review, I think I will make this throw instead of exit.

File: eng/common/scripts/Helpers/Package-Helpers.ps1
Comment (scbedd) [inline eng/common/scripts/Helpers/Package-Helpers.ps1:92]: :joy:

File: eng/common/scripts/Helpers/Package-Helpers.ps1
Comment (scbedd) [inline eng/common/scripts/Helpers/Package-Helpers.ps1:93]: Adding newline with a commit addressing any feedback

File: eng/common/scripts/Package-Properties.ps1
Comment (weshaggard) [inline eng/common/scripts/Package-Properties.ps1:124]: I assume these commented out lines can just be removed.

File: eng/common/scripts/Package-Properties.ps1
Comment (danieljurek) [inline eng/common/scripts/Package-Properties.ps1:110]: Multiple returns from the pipeline could trip this up. But the engsys would probably also be in a pretty messed up state if someone had multiple artifacts with the same name to begin with...

**Review by weshaggard — APPROVED** [review]

File: eng/common/scripts/Helpers/Package-Helpers.ps1
Comment (scbedd) [reply eng/common/scripts/Helpers/Package-Helpers.ps1:82]: `Install-ModuleIfNotInstalled` comes from it.

File: eng/common/scripts/Package-Properties.ps1
Comment (scbedd) [reply eng/common/scripts/Package-Properties.ps1:110]: Yeah if this happens it's going to hard error I believe. Because of the class usage the types actually matter, and if we try to convert an array to a HashTable I don't think that's going to work gracefully.

Let me try this with a deliberately broken artifact and let you know what happens.

File: eng/common/scripts/Package-Properties.ps1
Comment (scbedd) [reply eng/common/scripts/Package-Properties.ps1:124]: damnit I thought I cleaned everything. **shakes fist**

File: eng/common/scripts/Helpers/Package-Helpers.ps1
Comment (weshaggard) [reply eng/common/scripts/Helpers/Package-Helpers.ps1:82]: Oh sorry I miss read this I though you were importing Package-Helpers.ps1 not PSModule-Helpers.ps1.

We could probably shorten the relative path here though.

File: eng/common/scripts/Package-Properties.ps1
Comment (scbedd) [reply eng/common/scripts/Package-Properties.ps1:110]: My hunch of it being a hard error due to type conversion is correct!

image

```
Exception while parsing yml file C:\repo\azure-sdk-for-js\sdk\core\ci.yml: Cannot convert the "System.Object[]" value of type "System.Object[]" to type "System.Collections.Hashtable".
```

image

File: eng/common/scripts/Helpers/Package-Helpers.ps1
Comment (scbedd) [reply eng/common/scripts/Helpers/Package-Helpers.ps1:82]: Yeah definitely.

**Review by JimSuplizio — APPROVED** [review]
I'm approving this. The only concern I have is one that we'd talked about and that's not all languages have artifacts in their yml files. For example, neither Go nor C has them and I'm not currently sure what Rust is doing but they don't seem to have individual ci.yml files in sdk/ subdirectories yet...just something to be aware of.

**Comment by scbedd** [comment]
> I'm approving this. The only concern I have is one that we'd talked about and that's not all languages have artifacts in their yml files. For example, neither Go nor C has them and I'm not currently sure what Rust is doing but they don't seem to have individual ci.yml files in sdk/ subdirectories yet...just something to be aware of.

I will ensure that it works across all the repos before merging any of them 👍

**Pull request merged**
